File: java/leetcode/RandomPointCircle.java
//Leetcode problem 883 Generate Random Point in a Circle
//Solution written by Xuqiang Fang on 28 July, 2018
import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;
import java.util.HashSet;
import java.util.Set;
import java.util.PriorityQueue;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.Arrays;
//http://www.anderswallin.net/2009/05/uniform-random-points-in-a-circle-using-polar-coordinates/
class Solution {
    double radius, x_center, y_center;
    public Solution(double radius, double x_center, double y_center) {
        this.radius=radius;
        this.x_center=x_center;
        this.y_center=y_center;
    }
    
    public double[] randPoint() {
        double len= Math.sqrt(Math.random())*radius;
        double deg= Math.random()*360;
        double x= x_center+len*Math.cos(deg);
        double y= y_center+len*Math.sin(deg);
        return new double[]{x,y};
    }
}

public class RandomPointCircle{
	public static void main(String[] args){
		Solution s = new Solution();
	}
}

File: python/leetcode/RandomPointCircle.py
'''
Leetcode problem No 883 Generate Random Point in a Circle
Solution written by Xuqiang Fang on 28 July, 2018
'''
import random
import math
import matplotlib.pyplot as plt

class Solution(object):
    def __init__(self, radius, x_center, y_center):
        """
        :type radius: float
        :type x_center: float
        :type y_center: float
        """
        self.r = radius
        self.x_ = x_center
        self.y_ = y_center

    def randPoint(self):
        """
        :rtype: List[float]
        """
        l = math.sqrt(random.uniform(0,1))*self.r
        deg = random.uniform(0,1) * 360
        x = self.x_ + l * math.cos(deg)
        y = self.y_ + l * math.sin(deg)
        return [x, y]


def main():
    s = Solution(10, 5, -7.5)
    print(s.randPoint())
    print(s.randPoint())
    print(s.randPoint())
    print(s.randPoint())
    print(s.randPoint())
    ans_x = []
    ans_y = []
    for i in range(3000):
        a = s.randPoint()
        ans_x.append(a[0])
        ans_y.append(a[1])
    plt.scatter(ans_x, ans_y)
    plt.show()

if __name__ == "__main__":
    main()
